File: crates/core/examples/rpc.rs
use quicknode_sdk::{errors::SdkError, QuicknodeSdk, SdkFullConfig};

#[tokio::main]
#[allow(clippy::unwrap_used, clippy::expect_used)]
async fn main() {
    let config = SdkFullConfig::from_env().expect("Config from env failed");
    let qn = QuicknodeSdk::new(&config).expect("sdk failed to initialize");

    // Ensure Tooling Access is provisioned (idempotent; requires admin role).
    let status = qn
        .admin
        .tooling_access_status()
        .await
        .expect("tooling_access_status failed");
    println!("tooling access enabled: {}", status.enabled);
    if !status.enabled {
        match qn.admin.enable_tooling_access().await {
            Ok(s) => println!("enabled tooling access: {}", s.enabled),
            Err(e) => {
                eprintln!("could not enable tooling access: {e}");
                return;
            }
        }
    }

    // Make a JSON-RPC call. The SDK mints and refreshes the session JWT.
    match qn.rpc.call("eth_blockNumber", None, None).await {
        Ok(result) => println!("eth_blockNumber => {result}"),
        Err(e) => eprintln!("rpc call error: {e}"),
    }

    // Multichain: seed the per-network URL map from the endpoint id (returned by
    // status), then route a call to a specific network by its key.
    if let Some(id) = &status.endpoint_id {
        if let Ok(urls) = qn.admin.get_endpoint_urls(id).await {
            if let Some(mc) = urls.data.and_then(|d| d.multichain_urls) {
                qn.rpc
                    .set_networks(mc.into_iter().map(|(k, v)| (k, v.http_url)).collect());
                match qn
                    .rpc
                    .call("getSlot", None, Some("solana-mainnet".to_string()))
                    .await
                {
                    Ok(result) => println!("solana getSlot => {result}"),
                    Err(e) => eprintln!("solana rpc error: {e}"),
                }
            }
        }
    }

    // Demonstrate the typed JSON-RPC error path.
    match qn
        .rpc
        .call(
            "eth_getBalance",
            Some(serde_json::json!(["not-an-address"])),
            None,
        )
        .await
    {
        Ok(result) => println!("unexpected ok: {result}"),
        Err(SdkError::Rpc { code, message }) => {
            println!("got expected RpcError: code={code} message={message}");
        }
        Err(e) => eprintln!("other error: {e}"),
    }
}

File: npm/examples/rpc.ts
import { QuicknodeSdk, RpcError } from "../sdk";

async function main() {
  const qn = QuicknodeSdk.fromEnv();

  // Ensure Tooling Access is provisioned (idempotent; requires admin role).
  const status = await qn.admin.toolingAccessStatus();
  console.log(`tooling access enabled: ${status.enabled}`);
  if (!status.enabled) {
    try {
      const enabled = await qn.admin.enableToolingAccess();
      console.log(`enabled tooling access: ${enabled.enabled}`);
    } catch (e) {
      console.error(`could not enable tooling access: ${e}`);
      return;
    }
  }

  // Make a JSON-RPC call. The SDK mints and refreshes the session JWT.
  const blockNumber = await qn.rpc.call("eth_blockNumber");
  console.log(`eth_blockNumber => ${blockNumber}`);

  // Multichain: seed the per-network URL map (from the endpoint id in status),
  // then route a call to a specific network by its key.
  if (status.endpointId) {
    const urls = await qn.admin.getEndpointUrls(status.endpointId);
    if (urls.data?.multichainUrls) {
      const map = Object.fromEntries(
        Object.entries(urls.data.multichainUrls).map(([k, v]) => [k, v.httpUrl]),
      );
      qn.rpc.setNetworks(map);
      const slot = await qn.rpc.call("getSlot", [], "solana-mainnet");
      console.log(`solana getSlot => ${slot}`);
    }
  }

  // Demonstrate the typed JSON-RPC error path.
  try {
    await qn.rpc.call("eth_getBalance", ["not-an-address"]);
  } catch (e) {
    if (!(e instanceof RpcError)) throw e;
    console.log(`got expected RpcError: code=${e.code} message=${e.message}`);
  }

  // Custom endpoint URL: send a call to a fully-formed HTTP URL, bypassing the
  // Tooling Access endpoint and the session JWT entirely. Set it per-call here
  // (4th arg), or client-wide via `new RpcConfig({ endpointUrl })`.
  const customUrl = process.env.QN_RPC_ENDPOINT_URL;
  if (customUrl) {
    const result = await qn.rpc.call("eth_blockNumber", [], undefined, customUrl);
    console.log(`custom endpoint eth_blockNumber => ${result}`);
  }
}

main();
